File: Java/Beginner/k_backtracking/boj_14888.java
package Beginner.k_backtracking;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class boj_14888 {
    private static StringTokenizer st;
    private static int N;
    private static int max_val = (int) -(1e9 + 1);
    private static int min_val = (int) (1e9 + 1);
    private static final List<Integer> nums = new ArrayList<>();
    private static final List<Integer> ops = new ArrayList<>();
    private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

    private static void solve(int depth, int result) {
        if (depth == N) {
            max_val = Math.max(max_val, result);
            min_val = Math.min(min_val, result);
            return;
        }

        for (int i = 0; i < ops.size(); i++) {
            if (ops.get(i) > 0) {
                ops.set(i, ops.get(i) - 1);

                if (i == 0) solve(depth + 1, result + nums.get(depth));
                else if (i == 1) solve(depth + 1, result - nums.get(depth));
                else if (i == 2) solve(depth + 1, result * nums.get(depth));
                else if (i == 3) solve(depth + 1, result / nums.get(depth));

                ops.set(i, ops.get(i) + 1);
            }
        }
    }

    public static void main(String[] args) throws IOException {
        N = Integer.parseInt(br.readLine());

        st = new StringTokenizer(br.readLine());
        while (st.hasMoreTokens()) nums.add(Integer.parseInt(st.nextToken()));

        st = new StringTokenizer(br.readLine());
        while (st.hasMoreTokens()) ops.add(Integer.parseInt(st.nextToken()));

        solve(1, nums.get(0));
        System.out.println(max_val);
        System.out.println(min_val);
    }
}

File: Kotlin/src/main/kotlin/beginner/k_backtracking/boj_14888.kt
package beginner.k_backtracking

private var n = 0
private var maxVal = Int.MIN_VALUE
private var minVal = Int.MAX_VALUE
private lateinit var nums: IntArray
private lateinit var ops: IntArray

private fun dfs(depth: Int, result: Int) {
    if (depth == n) {
        maxVal = maxOf(maxVal, result)
        minVal = minOf(minVal, result)
        return
    }

    ops.forEachIndexed { i, op ->
        if (op > 0) {
            ops[i]--

            when (i) {
                0 -> dfs(depth + 1, result + nums[depth])
                1 -> dfs(depth + 1, result - nums[depth])
                2 -> dfs(depth + 1, result * nums[depth])
                3 -> dfs(depth + 1, result / nums[depth])
            }

            ops[i]++
        }
    }
}

fun boj_14888() = with(System.`in`.bufferedReader()) {
    n = readLine().toInt()
    nums = readLine().split(" ").map { it.toInt() }.toIntArray()
    ops = readLine().split(" ").map { it.toInt() }.toIntArray()

    dfs(1, nums[0])
    println(maxVal)
    println(minVal)
}